src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs

```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using NuGet.Common;
+
+namespace NuGet.Protocol.Plugins
+{
+    public static class PluginDiscoveryUtility
+    {
+        public static string InternalPluginDiscoveryRoot { get; set; } = null;
+
+        public static string GetInternalPlugins()
+        {
+            var rootDirectory = InternalPluginDiscoveryRoot ?? System.Reflection.Assembly.GetEntryAssembly()?.Location;;
+
+            return rootDirectory ?? Path.GetDirectoryName(rootDirectory);
+        }
+
+        public static string GetNuGetHomePluginsPath()
+        {
+            var nuGetHome = NuGetEnvironment.GetFolderPath(NuGetFolderPath.NuGetHome);
+
+            return Path.Combine(nuGetHome,
+                "plugins",
+#if IS_DESKTOP
+                "netfx"
+#else
+                "netcore"
+#endif
+                );
+        }
+
+        public static IEnumerable<string> GetConventionBasedPlugins(IEnumerable<string> directories)
+        {
+            var paths = new List<string>();
+            foreach (var directory in directories.Where(Directory.Exists))
```

Comment:
how is this list of directories being constructed?

can the responsibility of makign sure that the directory exists be delegated to the caller and you make this function safe by catching the appropriate exception ?

Reply:
It's up to the caller to construct the list.
Why shouldn't this function be safe itself?